### DISCUSSION TITLE
Java modules or automatic module names

### DISCUSSION THREAD

**Discussion by theigl** [resteasy/resteasy#4424] [General]
Hi all,

I recently started to adopt Java modules in my application. 

Maven currently complains very loudly about any libraries used that have no `module-info` or automatic module name. I was quite surprised that Resteasy is one of the libraries that has neither:

```
[WARNING] * Required filename-based automodules detected: 
 resteasy-core-6.2.11.Final.jar, 
 resteasy-core-spi-6.2.11.Final.jar, 
 resteasy-validator-provider-6.2.11.Final.jar, 
 resteasy-multipart-provider-6.2.11.Final.jar
```

My questions are:

- Is there a reason for this? 
- Are there plans to fully modularize Resteasy? 
- Or would you accept a PR that adds automatic module names?

**Comment by jamezp** [comment]
The biggest issue RESTEasy faces is it needs all the dependencies it uses to be modularized as well. It has been quite a while since I've looked, but there are a lot of the dependencies that at the time were not modules. For example, the Apache HTTP Client. There could be others now too, but we need to analyze each dependency to determine this.

**Reply by theigl** [reply]
This is only true if you want your users to be able to build stripped down JREs with JLink. If you simply want to modularize Resteasy so you can benefit from improved encapsulation, then you could do it now. Apache Wicket for instance is fully modularized, but has several dependencies that are not modularized.

Having said that, modularization of a large project like Resteasy is likely going to be difficult. In the meantime, I suggest that you add automatic module names to all your JARs. I can provide a PR if you want.

**Reply by jamezp** [reply]
I'd really like to analyze the dependencies at some point. It might be for now that automatic module names is the only way to go. I believe this wad done at some point and reverted for some reason. I don't recall the details off the top of my head though.

The hardest part will be creating a testing environment I think. The integration tests won't work because use WildFly which has it's ow module system. That said, testing with the https://github.com/resteasy/resteasy-junit-extension might work. However, I'd need to make that a JPMS module which should be doable.

**Reply by theigl** [reply]
> I'd really like to analyze the dependencies at some point. 

That definitely makes sense.

> It might be for now that automatic module names is the only way to go. I believe this wad done at some point and reverted for some reason. I don't recall the details off the top of my head though.

I went through the complete history of the resteasy-core POM but I found no trace of an automatic module name. As I said, I can provide a PR for automatic module names (it's basically just a single configuration parameter for the maven-jar-plugin), but it would be good to know if this really was reverted in the past.

What I found looking through the POM's history is this PR #1697. It looks like the work done to support WildFly modules already laid most of the ground work for further modularization.

**Reply by theit** [reply]
Any news wrt automatic module names and/or `module-info`?

**Reply by jamezp** [reply]
Nothing great to share yet. I'm hoping very soon to be able to look at this. The first step is analyzing what dependencies don't have module names yet, then I can come up with a plan to do something.

**Reply by jamezp** [reply]
As an FYI, I've got a draft up now https://github.com/resteasy/resteasy/pull/4903. I do plan on back-porting to 6.2 as well.

**Reply by jamezp** [reply]
This has now been done. Not quite as soon as I wanted, but it's done and will be in the 6.2.17.Final and 7.0.3.Final releases :)

https://github.com/resteasy/resteasy/pull/4903
https://github.com/resteasy/resteasy/pull/4908

**Reply by theigl** [reply]
🚀 Fantastic. Thanks @jamezp
